evm_arithmetization/src/generation/mod.rs

```diff
@@ -309,39 +312,220 @@ pub fn generate_traces<F: RichField + Extendable<D>, const D: usize>(
     Ok((tables, public_values))
 }
 
-fn simulate_cpu<F: Field>(state: &mut GenerationState<F>) -> anyhow::Result<()> {
-    let halt_pc = KERNEL.global_labels["halt"];
+/// A State is either an `Interpreter` (used for tests and jumpdest analysis) or
+/// a `GenerationState`.
+pub(crate) enum State<'a, F: Field> {
```

Comment:
Why don't you create a trait `State` with all this functions and implement the trait for `GenerationState` and `Interpreter`? In that case you will benefit from the static dispatch.

Reply:
And maybe it fits better on state.rs